Add Trello integration

ScottG489 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Initial integration with Trello. In the interest of keeping this PR as small as possible, scope was limited to allowing you to track the number of cards in the lists on a Trello board. Meaning, an entity represents a list on a given board and it's value is the number of cards in that list. Future plans include services for card creation/update, better auth (I believe [OAuth is supported](https://developer.atlassian.com/cloud/trello/guides/rest-api/authorization/#using-basic-oauth)), and shifting it's IoT class to cloud push via [webhooks](https://developer.atlassian.com/cloud/trello/guides/rest-api/webhooks/).

I have tested this out personally with two separate Trello accounts, one paid and one free.

We're using the latest version of [py-trello](https://github.com/sarumont/py-trello) for external communication.

Otherwise, I haven't written Python in a while (though I'm a FT SWE), so on top of the usual comments on fixes required for merging, I'm also looking for suggestions on how I can improve so my next PR (or this one) can be even better :)

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/27526

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

ScottG489 committed:
Initialize new Trello integration

ScottG489 committed:
Initial Trello integration implementation

User enters credentials via config flow then selects 0-n boards they
would like to have card counts for each list on that board. There is
also an options flow where the user can add or remove boards freely.

For sensor updates, we not only use a coordinator to fetch all data at
once, we use Trello's REST API /batch endpoint to do so with a single
request. This is particularly important because of Trello's rate limits.

ScottG489 committed:
Add tests for new Trello integration

ScottG489 commented:
I see we're a little short on coverage. The missing coverage is mostly various one-line filler code or constructors. Is this a hard requirement? Regardless, I'll try to spend some time to look over the tests and get the coverage up.

ScottG489 committed:
Add module to strict typing

This also includes changes necessary to fulfill the strict-typing
requirements by adding type hints where necessary.

joostlek commented on homeassistant/components/trello/config_flow.py:
```diff
@@ -0,0 +1,220 @@
+        self.ids_boards: dict[str, dict[str, str]] = {}
+        self.trello_adapter: TrelloAdapter
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(
+        config_entry: config_entries.ConfigEntry,
+    ) -> config_entries.OptionsFlow:
+        """Create the options flow."""
+        return OptionsFlowHandler(config_entry)
```
Perhaps move the options flow for a later PR

ScottG489 replied:
That's a good point, should have thought of that. That will clean up a lot of code and tests. Think I might do that actually!

joostlek commented on homeassistant/components/trello/config_flow.py:
```diff
@@ -0,0 +1,220 @@
+
+    VERSION = 1
+
+    def __init__(self) -> None:
+        """Initialize config flow."""
+        self.api_key: str = ""
+        self.api_token: str = ""
+        self.user_email: str = ""
+        self.user_id: str = ""
+        self.ids_boards: dict[str, dict[str, str]] = {}
+        self.trello_adapter: TrelloAdapter
```
Putting everything in a dict solves the problem of having to initialize every value

ScottG489 replied:
The big drawback of using a `dict` here is that you can't plainly see what information the config flow is gathering to use to create the config entry.

This could be cleaned up with a data object, but the values would still need to be initialized. In it's current state, I think there are few enough members where that abstraction isn't necessary quite yet, though.

joostlek commented on homeassistant/components/trello/config_flow.py:
```diff
@@ -0,0 +1,220 @@
+
+        return await self._show_board_form()
+
+    async def async_step_finish(self, user_input: dict[str, Any]) -> FlowResult:
```
Perhaps use another name for this step, finish is a bit too generic imo, maybe boards is more recognizable

joostlek commented on homeassistant/components/trello/config_flow.py:
```diff
@@ -0,0 +1,220 @@
+            step_id="user", data_schema=CREDS_FORM_SCHEMA, last_step=False
+        )
+
+    async def _show_board_form(self) -> FlowResult:
+        return self.async_show_form(
+            step_id="finish",
+            data_schema=_get_board_select_schema(self.ids_boards),
+            last_step=True,
+        )
```
I'd suggest getting rid of the separate functions for showing forms. It looks confusing for new contributors and you always have to trace back to the place where it was called. (for example what if this method is called while ids_boards is None for some reason. It can get confusing)

ScottG489 replied:
I can see where you're coming from with having to jump back and forth. I think it's more important to keep the `async_step` methods as readable as possible though. That's where the high level logic is and all the form definitions makes it much more difficult to grok. Also having all the form definitions encapsulated in these methods makes it easier to reference.

As a new contributor myself, it was really difficult to read through a lot of config flows for reference because they didn't abstract away a lot of higher level behavior.

I'm with you there on the `self.ids_boards`  reference though. To make it easier to reason about, I'm going to pass that in as a param.

joostlek replied:
I agree with you, in one of my first PRs I also did this and then I got this as feedback (https://github.com/home-assistant/core/pull/91601#discussion_r1169927319)

ScottG489 replied:
Thanks, that's good to know. I still feel pretty strongly that this should be abstracted away into a method for readability in the `async_step` method. 

Otherwise, it does seem like `add_suggested_values_to_schema` would be appropriate to use here. However, looking at it again, I actually don't need to set defaults anymore since I've removed the options flow. So I'm removing the logic that would make `add_suggested_values_to_schema` necessary, but I'll definitely implement it in the follow up PR.

joostlek commented on homeassistant/components/trello/config_flow.py:
```diff
@@ -0,0 +1,220 @@
+        :param user_input: None initially or users api_key and api_token.
+        :return:
+        """
+        if user_input is None:
+            return await self._show_creds_form()
+
+        self.api_key = user_input[CONF_API_KEY]
+        self.api_token = user_input[CONF_API_TOKEN]
+        self.trello_adapter = _create_trello_adapter(
+            user_input[CONF_API_KEY], user_input[CONF_API_TOKEN]
+        )
+
+        try:
+            member = await self._get_current_member()
+        except Unauthorized as ex:
+            return await self._show_error_creds_form(ex)
+
+        self.user_id = member.id
+        self.user_email = member.email
+        self.ids_boards = await self._fetch_all_boards()
+
+        return await self._show_board_form()
```
Not sure if GitHub orders all my comments like they are added, but to get rid of these small functions you can invert this method. Only if user input is filled go into the if section. https://github.com/home-assistant/core/blob/dev/homeassistant/components/lastfm/config_flow.py is an example.

joostlek commented on homeassistant/components/trello/config_flow.py:
```diff
@@ -0,0 +1,220 @@
+
+    async def _show_creds_form(self) -> FlowResult:
+        return self.async_show_form(
+            step_id="user", data_schema=CREDS_FORM_SCHEMA, last_step=False
```
I think last_step=False is the default, so you can omit that.

ScottG489 replied:
It looks like it's actually the other way around, because removing `last_step=false` causes the buttons to say "submit" instead of "next". I also wonder if there's other functionality that uses this value.

For now I'll remove `last_step=true` unless someone can point out how it isn't redundant and used elsewhere.

joostlek replied:
Hmm, I can't remember, but I could also be wrong here

joostlek commented on homeassistant/components/trello/config_flow.py:
```diff
@@ -0,0 +1,220 @@
+        return await self.hass.async_add_executor_job(self.trello_adapter.get_member)
+
+
+class OptionsFlowHandler(config_entries.OptionsFlow):
```
I'd suggest moving this to a separate PR. Also, if you're continuing to work on it in the meantime, checkout OptionsFlowWithConfigEntry (check the lastfm one how it is used)

joostlek commented on homeassistant/components/trello/config_flow.py:
```diff
@@ -0,0 +1,220 @@
+    return TrelloAdapter(TrelloClient(api_key=api_key, api_secret=api_token))
+
+
+def _get_board_select_schema(
+    boards: dict[str, dict], default: list[str] | None = None
+) -> Schema:
+    if default is None:
+        default = []
+    options = {key: value["name"] for key, value in boards.items()}
+    return vol.Schema(
+        {vol.Required(USER_INPUT_BOARD_IDS, default=default): cv.multi_select(options)}
+    )
```
I am not sure if this should be moved inside both of the flows. I think it would be useful, because then you can use `self.add_suggested_values_to_schema(` to add the defaults

ScottG489 replied:
I ended up removing the defaulting behavior since it was only needed for the options flow so that won't be needed for now.

joostlek commented on homeassistant/components/trello/const.py:
```diff
@@ -0,0 +1,15 @@
+"""Constants for the trello_int integration."""
```
```suggestion
"""Constants for the trello integration."""
```

joostlek commented on homeassistant/components/trello/const.py:
```diff
@@ -0,0 +1,15 @@
+
+LOGGER = logging.getLogger(__package__)
+
+CONF_API_TOKEN = "api_token"
```
Euh, maybe checkout a better name for api_token, in py-trello they are talking about api key, api secret, (oauth) token and (oauth) secret. so seeing api_token and api_key used both looks confusing

ScottG489 replied:
Yeah, thinking about it now it might actually be an OAuth token. However, it seems weird that they would have you get an OAuth token manually, but maybe that isn't completely out of the ordinary.

I did go with the name "token", though, because that's what Trello themselves called it.

joostlek replied:
yea but api_token api_key look similair, better make it normal token then

ScottG489 replied:
"API Key" and "API Token" are both preexisting naming in [`strings.json`](https://github.com/home-assistant/core/blob/5a18de67e5593cb965f3f343a0dfbbcde6b427e5/homeassistant/strings.json#L46-L47). Also "token" by itself could have other meanings, so I'd like to be explicit with what it's a token for.

joostlek commented on homeassistant/components/trello/coordinator.py:
```diff
@@ -0,0 +1,69 @@
+"""Data update coordinator for the Steam integration."""
```
```suggestion
"""Data update coordinator for the Trello integration."""
```

ScottG489 replied:
Busted :( Nice catch.

joostlek replied:
In dutch we have a saying "Better stolen well than invented poorly"

joostlek commented on homeassistant/components/trello/const.py:
```diff
@@ -0,0 +1,15 @@
+
+CONF_OPTIONS_BOARDS = "boards"
+
+USER_INPUT_BOARD_IDS = "board_ids"
```
```suggestion
CONF_BOARD_IDS = "board_ids"
```
I think this prefix works better. I see that you're using this only as temporary, to fetch the boards but I think you can just use 1 variable

ScottG489 replied:
Yeah I'm kinda split on this. Although it's tempting to reduce duplication by reusing strings, the strings are really representing keys for different things. Though I'm not consistent with this myself since I use other `CONF_*` consts as keys on `user_input` [elsewhere](https://github.com/home-assistant/core/pull/93580/files#diff-a4883022abef83076d53d668dbc99a6a66fe90a152e0381b9ac04f818d05b729R57-R58).

joostlek commented on homeassistant/components/trello/coordinator.py:
```diff
@@ -0,0 +1,69 @@
+            lists = board_lists_response["200"]
+            for list_ in lists:
+                list_card_counts[list_["id"]] = len(list_["cards"])
+        else:
+            LOGGER.error(
+                "Unable to fetch lists for board with ID '%s'. Response was: %s)",
+                board_ids[i],
+                board_lists_response,
+            )
+            continue
```
Not sure yet how the best way to handle it is, but if it fails, the sensor would still show the old data instead of becoming unavailable

ScottG489 replied:
Perhaps something like [this from the steam_online](https://github.com/home-assistant/core/blob/147636357e1cff3c96ea278f2b0bdb6b272f55cb/homeassistant/components/steam_online/coordinator.py#L69-L72) integration?

ScottG489 replied:
So there's one potential problem with this implementation. We process batch requests here and if we throw an error like `UpdateFailed`, then one failed batch will cause a complete failure in updating all sensors. This might be the correct action to take since there shouldn't be an error here and maybe it should be expected for the entire integration to error out until the problem is corrected.

Also in this context it's difficult to signal which boards and lists are affected by a single request in the batch failing since this method only returns a list of Trello lists.

ScottG489 replied:
This has been resolved with new changes to the coordinator which now provides richer data that the sensor can use to determine if it should mark itself as unavailable as a result of it's associated request in the coordinator failing.

joostlek commented on homeassistant/components/trello/strings.json:
```diff
@@ -0,0 +1,33 @@
+    "step": {
+      "user": {
+        "data": {
+          "api_key": "API Key",
+          "api_token": "Token"
```
I think there are lokalize strings for these values, please check them and add them here

joostlek reviewed (commented):
Right, I am giving feedback on some things I got feedback on in the past. Let me know if you need clarification or examples :). I think it would be better to directly get into oauth, since it works a lot better than having to think of a way to migrate users. (or there is a simple thing for this I don't know yet)

ScottG489 committed:
Remove Trello options flow and associated tests

It's extraneous for the initial PR and adds a lot of complexity. Will be
added back in a follow up PR.

ScottG489 committed:
Various refactoring fixes based on PR comments

- Replace member reference with new parameter to make method easier to
  reason about
- Rename flow step to better describe what it does
- last_step=True is redundant because by default the button will say
  "submit"
- Fix copy/paste problems with various docstrings
- Remove redundant TrelloEntity and move attributes to TrelloSensor
  subclass

ScottG489 committed:
Remove logic to set preselected boards when prompting the user

Remove logic to set a default value when presenting a list of boards to
select. There will never be preselected boards when prompting the user
to select them since the options flow has been removed.

ScottG489 committed:
Refactor initial config flow step to be it's own method

This separates concerns so that each flow step method is responsible for
processing one form.

ScottG489 committed:
Reference common string translations for config flow creds labels

Note that the "Token" label is changed to "API Token".

ScottG489 committed:
Rename constant USER_INPUT_BOARD_IDS to CONF_BOARD_IDS

So that we're more consistent with key names for constants

ScottG489 commented:
Unfortunately it looks like I might not be able to implement OAuth at all since [Trello only supports OAuth 1.0](https://developer.atlassian.com/cloud/trello/guides/rest-api/authorization/#using-basic-oauth) and I think we only [support OAuth 2.0](https://developers.home-assistant.io/docs/config_entries_config_flow_handler/#configuration-via-oauth2).

Can anyone confirm?

I'm chatting with their support to see if they have any plans to implement OAuth2. They already mentioned they're doing some changes to auth.

ScottG489 committed:
Update coordinator data every 60s

The coordinator was only updating on startup and it should be polling
for state changes.

I mistakenly thought 60s was the default when not specified.

ScottG489 commented:
Regarding Trello supporting OAuth2 - support got back to me:

> [engineering does] have a project to make the Trello API support OAuth 2.0, it hasn't been scheduled yet, so we can't say for sure how soon that will be. It will happen though — eventually.

So OAuth2 is likely not going to happen with this PR, but as soon as it's available I'll definitely implement it in a follow up PR.

joostlek commented:
> Doesn't look related to my changes and it passes for me locally.
> 
> I took a look at it, but didn't see any recent changes to that test method. I also tried taking a look at recent recorder changes, but I'm afraid I'm not really familiar enough.

These sometimes fail, I am also not familiar so unless your tests are failing I wouldn't mind them. (I don't think when editing an integration this would fail, so unless you're actually working on core stuff, don't bother)

joostlek commented on homeassistant/components/trello/config_flow.py:
```diff
@@ -0,0 +1,141 @@
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Prompt user for Trello API credentials."""
+        return await self._show_creds_form()
```
You're now always moving the user to another step. I think it's better to move that creds step to the user step so it stays in line with other integrations.

ScottG489 replied:
I wouldn't really say I'm moving the user to another step. This is simply describing what we're doing to display the initial dialog. I also discuss my issues with the alternative in a [previous comment](https://github.com/home-assistant/core/pull/93580#discussion_r1209449037).

I can tell you as a new developer, looking at the different flow step functions of other integrations, it was really confusing. Using the `user_input is None` pattern has a lot of downsides:
- It was confusing why a step was calling itself. Especially as part of the happy path. Felt very unnatural.
- Makes the function necessarily large by doing too much.
- It doesn't take advantage of the first class concept of "steps" that HA has for config flows. Otherwise the "user" step is really rendering multiple steps.

I think it's fine to stay in line with other integrations and patterns used when they make sense, but I think `user_input is None` is something that needs to be replaced.

edenhaus replied:
We discussed this internally, and we want to achieve the same pattern throughout the code base.
Similarity makes it easier to review and maintain the code.

Please adopt the config flow so the code of the steps is similar to the other integrations.

joostlek commented on homeassistant/components/trello/config_flow.py:
```diff
@@ -0,0 +1,141 @@
+
+        :param user_input: User's selected boards
+        """
+        boards: dict[str, dict] = await self._get_boards_lists(
+            user_input[CONF_BOARD_IDS]
+        )
```
So if I understand correctly, you're storing the board id and the list ids at setup?

What if the lists change? If so, check if you can move the fetching of the list ids to some other place like at setup or every update, depending on the API speed, stability and rate limits.

ScottG489 replied:
Yes this is a great point and something I had been considering. I think with the initial implementation I had decided to forego accounting for some dynamic aspects such as handling board or list name changes or additions or deletions of boards or lists. Primarily because it seemed like extraneous functionality and I wanted to keep the PR as small as possible.

However, I'm also not sure of the best way to handle some of these situations. For instance, should additional lists just be added to the device if a new board is added or should this be something that should first require manual intervention from the user such as reloading the integration or going through the options flow?

joostlek replied:
I think storing data with the data from the API forces you in a position that you will need to make a migration when you decide you want to get that data dynamically (as your data structure changes in the config entry). I think it's better to move everything to dynamically load. This way you don't have to bother with forcing an options flow etc. I am also looking into a way to maybe slow down some factors of a sensor update. (For YouTube for example, a channel name doesn't change every day, so fetching this data would be overhead.)

ScottG489 replied:
Great suggestions. I've refactored the coordinator to provide richer data the sensor can use to keep itself updated at runtime (e.g. if the lists name changes or an update fails). I've also removed all data from the config entry except the board IDs that the user selected and when setting up the sensors they use the up-to-date coordinator data instead of static config entry data.

joostlek commented on homeassistant/components/trello/config_flow.py:
```diff
@@ -0,0 +1,141 @@
+            data_schema=_get_board_select_schema(ids_boards),
+        )
+
+    async def _show_error_creds_form(self, ex: Unauthorized) -> FlowResult:
```
I'd still say move this back to the step forms. Also, can there be other exceptions thrown like timeout? (Not saying every integration has support for this, but now you're only expecting Unauthorized)

ScottG489 replied:
I think the [reasoning in my other comment](https://github.com/home-assistant/core/pull/93580#discussion_r1208829428) still holds where it's more important to keep the step functions as readable as possible.

Regarding only catching a specific exception, I'm not sure the best way to handle this here. I'm somewhat intentionally only catching this one error because for the broad range of other errors that could happen here, I don't know how to handle those. So if something else is thrown, the error will bubble up and the config flow dialog will display "Unknown error occurred" which I think is appropriate.

What do you think?

joostlek commented on homeassistant/components/trello/config_flow.py:
```diff
@@ -0,0 +1,141 @@
+        return await self.hass.async_add_executor_job(self.trello_adapter.get_member)
+
+
+def _create_trello_adapter(api_key: str, api_token: str) -> TrelloAdapter:
+    return TrelloAdapter(TrelloClient(api_key=api_key, api_secret=api_token))
```
Don't see a lot of pros of having this separated

ScottG489 replied:
Same. I think it was left over from when I still had the options flow. I do think there's merit in having a helper method to create a `TrelloAdapter` from creds though. What do you think of a `TrelloAdapter.from_creds(key,  token)` `@classmethod`?

joostlek commented on homeassistant/components/trello/const.py:
```diff
@@ -0,0 +1,14 @@
+CONF_USER_EMAIL = "user_email"
+CONF_BOARD_IDS = "board_ids"
+
+CONF_OPTIONS_BOARDS = "boards"
```
CONF and OPTIONS sound a bit the same, how does CONF_BOARDS sound?

ScottG489 replied:
I think I was trying to differentiate keys for options and config data, but not sure there's any good reason to do that. So I think this change would make sense.

joostlek commented on homeassistant/components/trello/sensor.py:
```diff
@@ -0,0 +1,83 @@
+        self._attr_name = _list["name"]
+        self._attr_has_entity_name = True
+
+    def update(self) -> None:
+        """Set the card count for the sensor's list."""
+        self._attr_native_value = self.coordinator.data[self.list_id]
```
You don't need to set _attr_native_value if youre implementing native_value and returning the same thing

ScottG489 replied:
I swear I tried removing that and things didn't work. Will revisit.

joostlek commented on homeassistant/components/trello/strings.json:
```diff
@@ -0,0 +1,25 @@
+      },
+      "boards": {
+        "data": {
+          "board_ids": "Board names"
```
I'd go with "Boards" as you're selecting boards and not board names.  (You're showing board names, but you're not selecting them ;))

ScottG489 replied:
Makes sense. Will make this change.

ScottG489 committed:
Rename constant CONF_OPTIONS_BOARDS to CONF_BOARDS

ScottG489 committed:
Prompt user to select "Boards" not "Board names"

The user isn't choosing board names but rather selecting the boards
themselves they want to set up within the integration, so this wording
is more accurate.

ScottG489 committed:
Move helper method to be a factory class method

Move a helper method for creating a TrelloAdapter from creds to be a
factory class method instead.

ScottG489 committed:
Remove unused sensor update method

This method isn't needed when the native_value method is also being
used.

ScottG489 committed:
Set entities unavailable if there was a problem fetching their data

When there was a problem fetching data for a particular board, we had no
way to indicate to the sensors that there was a problem since we only
retained card count information in the coordinator data.

This change stores a richer set of data in the coordinator so the sensor
can mark itself as unavailable when appropriate.

- Use data from coordinator when setting up sensors rather than data
  stored in config entry. The config entry will soon be slimmed down to
  only contain configured board ids

ScottG489 committed:
Change friendly name of list entity when changed in remote source

If the name of a Trello list was updated in Trello itself, we were not
reflecting this information the name locally. Now that we have richer
data being stored in the coordinator, we have access to this and when
the sensor updates so will the name.

ScottG489 committed:
Only save selected board ids into config entry

Previously we were saving board and list names in the config and using
this information to set up entities. However, some of this information
can change at run time.

Instead, we get this information from the coordinator and only save the
IDs of the boards the user selected.

ScottG489 committed:
Replace string usages for config data keys with constants

ScottG489 committed:
Break out sensor availabilty and name calculation from native_value

I was advised that native_value should not be responsible for handling
the the calculation of these other values, so I've broken them out into
their appropriate methods.

ScottG489 committed:
Replace const defined for integration with preexisting global const

No need for us to define and use our own CONF_API_TOKEN const when this
const is already defined in the project-wide const file.

ScottG489 committed:
Make trello sensor tests more comprehensive

The sensor test is more of an integration test now where it tests the
full setup of the component and asserts on states actually pulled from
the state machine (hass). It also tests updates to the sensors and
ensures errors returned from the trello client are handled properly by
marking the corresponding sensor as unavailable.

This also makes the coordinator tests obsolete, so they have been
removed.

ScottG489 committed:
Merge branch 'dev' into trello_integration

ScottG489 committed:
Fix ruff issues

ScottG489 committed:
Merge branch 'dev' into trello_integration

ScottG489 committed:
Fix mypy issue

I believe this issue is due to the fact that the class DeviceInfo moved
packages.

ScottG489 committed:
Merge branch 'dev' into trello_integration

joostlek commented on homeassistant/components/trello/__init__.py:
```diff
@@ -0,0 +1,58 @@
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up from a config entry."""
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
```
Let's do the coordinator initializing in the init file

joostlek commented on homeassistant/components/trello/__init__.py:
```diff
@@ -0,0 +1,58 @@
+class TrelloAdapter:
+    """Adapter for Trello lib's client."""
+
+    def __init__(self, client: TrelloClient) -> None:
```
Make hass a parameter and do all the `self.hass.async_add_executor_job` in here and have an async API

ScottG489 replied:
We can discuss further if you feel very strongly about this, but I'm going to keep this as is. The adapter is meant to be a very thin, local client that wraps the 3rd party dependency for the purposes of this integration. Having only the 3rd party client as a dependency keeps it simple and completely unit testable. It also gives us a place to put a well defined interface with the 3rd party client.

Currently there are some other places where we interact with the 3rd party client directly, but eventually I'd like to refactor all usages to go through this adapter instead.

joostlek commented on homeassistant/components/trello/sensor.py:
```diff
@@ -0,0 +1,97 @@
+class TrelloSensor(CoordinatorEntity[TrelloDataUpdateCoordinator], SensorEntity):
+    """Representation of a TrelloSensor."""
+
+    _attr_native_unit_of_measurement: str | None = "Cards"
+    _attr_state_class: SensorStateClass | str | None = SensorStateClass.MEASUREMENT
```
```suggestion
    _attr_native_unit_of_measurement = "Cards"
    _attr_state_class = SensorStateClass.MEASUREMENT
```

joostlek commented on homeassistant/components/trello/sensor.py:
```diff
@@ -0,0 +1,97 @@
+        self.coordinator = coordinator
+        self._attr_unique_id = f"list_{self.list_id}".lower()
+        self._attr_name = list_.name
+        self._attr_has_entity_name = True
```
can be shorthand

joostlek commented on homeassistant/components/trello/sensor.py:
```diff
@@ -0,0 +1,97 @@
+        super().__init__(coordinator)
+        self.board = board
+        self.list_id = list_.id
+        self.coordinator = coordinator
```
I think this is already done in CoordinatorEntity

joostlek commented on homeassistant/components/trello/sensor.py:
```diff
@@ -0,0 +1,97 @@
+        self.board = board
+        self.list_id = list_.id
+        self.coordinator = coordinator
+        self._attr_unique_id = f"list_{self.list_id}".lower()
```
```suggestion
        self._attr_unique_id = f"list_{list_.id}".lower()
```

joostlek commented on homeassistant/components/trello/sensor.py:
```diff
@@ -0,0 +1,97 @@
+        super()._handle_coordinator_update()
+
+    @property
+    def device_info(self) -> DeviceInfo:
```
Can be shorthand

ScottG489 committed:
Remove unnecessary type annotations

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

ScottG489 committed:
Reference locally scoped var rather than member in constructor

Easier to reason about if we're referencing a var scoped locally.

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

ScottG489 committed:
Use attr shorthand where possible

Not necessary to break it out into a property method since we don't
require any extra logic to create the DeviceInfo.

ScottG489 committed:
This declaration is already done in BaseCoordinatorEntity

ScottG489 committed:
Don't check if there are no boards configured

Currently it isn't possible to create a config entry without boards.
Removing this code so it isn't confusing.

ScottG489 committed:
Initialize coordinator at component level

The coordinator is shared between all sensors so it makes more sense to
initialize it there.

- Refactor mock config entry setup for tests.

edenhaus commented on homeassistant/components/trello/sensor.py:
```diff
@@ -0,0 +1,82 @@
+    @property
+    def available(self) -> bool:
+        """Determine if sensor is available."""
+        board = self.coordinator.data[self.board.id]
```
Please call the parent available function too

edenhaus replied:
```suggestion
	if not super.available():
	    return False
        board = self.coordinator.data[self.board.id]
```

We need to call it `super.available()` otherwise it could happen that we get a `KeyError` on accessing `self.coordinator.data`

edenhaus commented on homeassistant/components/trello/coordinator.py:
```diff
@@ -0,0 +1,87 @@
+        """Fetch data for all sensors using batch API endpoint."""
+        board_lists_query_params = "fields=name"
+        list_cards_query_params = "fields=name&cards=open&card_fields=idCard"
+        urls = ",".join(
+            f"/boards/{board_id}?{board_lists_query_params},/boards/{board_id}/lists?{list_cards_query_params}"
+            for board_id in self.board_ids
+        )
```
This looks for me implementation details, which should go into the library

ScottG489 replied:
I've pulled the implementation details of the batch endpoint into a new Trello lib and refactored them out from this PR. Thanks for pointing this out!

ScottG489 committed:
Merge branch 'dev' into trello_integration

ScottG489 committed:
Hassfest fixes

When resolving conflicts from master in mypy.ini and .strict-typing I
accidentally got the alphabetical ordering wrong.

ScottG489 committed:
Merge branch 'dev' into trello_integration

ScottG489 committed:
Merge branch 'dev' into trello_integration

ScottG489 committed:
Merge branch 'dev' into trello_integration

ScottG489 committed:
Fix linting issue in sensor test

Fixes the following lint error:
W7421: Relative import should be used (hass-relative-import)

ScottG489 committed:
Use new trello client lib with native support for batch requests

ha-py-trello has native support for batch requests and py-trello
doesn't.

Refactor coordinator to use the new client. This removes implementation
details from HA as they are now in the lib.

ScottG489 committed:
Consider sensor's parent availability

When determining sensor's availability we should consider it's super
classes availability.

ScottG489 committed:
Merge branch 'dev' into trello_integration

ScottG489 committed:
Refactor config flow steps to be consistent with other integrations

ScottG489 committed:
Merge branch 'dev' into trello_integration

ScottG489 committed:
Merge branch 'dev' into trello_integration

edenhaus commented on homeassistant/components/trello/config_flow.py:
```diff
@@ -0,0 +1,118 @@
+    def _get_config_data(self) -> dict[str, str]:
+        return {
+            CONF_API_KEY: self.api_key,
+            CONF_API_TOKEN: self.api_token,
+            CONF_USER_ID: self.user_id,
+            CONF_USER_EMAIL: self.user_email,
+        }
+
+    async def _show_creds_form(self) -> FlowResult:
+        return self.async_show_form(
+            step_id="user", data_schema=CREDS_FORM_SCHEMA, last_step=False
+        )
+
+    async def _show_board_form(
+        self, ids_boards: dict[str, dict[str, str]]
+    ) -> FlowResult:
+        return self.async_show_form(
+            step_id="boards",
+            data_schema=_get_board_select_schema(ids_boards),
+        )
+
+    async def _show_error_creds_form(self, ex: Unauthorized) -> FlowResult:
+        LOGGER.error("Unauthorized: %s)", ex)
+        return self.async_show_form(
+            step_id="user",
+            data_schema=CREDS_FORM_SCHEMA,
+            errors={"base": "invalid_auth"},
+            last_step=False,
+        )
+
+    async def _fetch_all_boards(self) -> dict[str, dict[str, str]]:
+        return await self.hass.async_add_executor_job(self.trello_adapter.get_boards)
+
+    async def _get_current_member(self) -> Member:
+        return await self.hass.async_add_executor_job(self.trello_adapter.get_member)
+
+
+def _get_board_select_schema(boards: dict[str, dict]) -> Schema:
+    options = {key: value["name"] for key, value in boards.items()}
+    return vol.Schema({vol.Required(CONF_BOARD_IDS): cv.multi_select(options)})
```
```suggestion
USER_FORM_SCHEMA = vol.Schema(
    {
        vol.Required(CONF_API_KEY): str,
        vol.Required(CONF_API_TOKEN): str,
    }
)


class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
    """Handle a config flow for the initial setup of a Trello integration."""

    VERSION = 1

    def __init__(self) -> None:
        """Initialize config flow."""
        self._data = {}
        self._trello_adapter: TrelloAdapter
        self._ids_boards: dict[str, dict[str, str]] = {}


    async def async_step_user(
        self, user_input: dict[str, Any] | None = None
    ) -> FlowResult:
        """Prompt user for Trello API credentials."""
        errors = {}
        if user_input:
            self._trello_adapter = TrelloAdapter.from_data(user_input)
            try:
                member = await self.hass.async_add_executor_job(self._trello_adapter.get_member)
            except Unauthorized as ex:
                LOGGER.exception("Unauthorized", ex)
                errors={"base": "invalid_auth"}
            else:
                self._data = {
                    **user_input,
                    CONF_USER_ID: member.id,
                    CONF_USER_EMAIL: member.email,
                }

                await self.async_set_unique_id(member.id)
                self._abort_if_unique_id_configured()

                self._ids_boards = await self.hass.async_add_executor_job(self._trello_adapter.get_boards)
                return await self.async_step_boards()

        return self.async_show_form(
            step_id="user", data_schema=USER_FORM_SCHEMA,errors=errors, last_step=False
        )

    async def async_step_boards(self, user_input: dict[str, Any]) -> FlowResult:
        """Select desired boards to have card counts of per list.

        :param user_input: User's selected boards
        """
        if user_input:
            return self.async_create_entry(
                title=self.user_email, data=self._data, options=user_input
            )

        options = {key: value["name"] for key, value in self._ids_boards.items()}
        return self.async_show_form(
            step_id="boards",
            data_schema=vol.Schema({vol.Required(CONF_BOARD_IDS): cv.multi_select(options)}),
        )
```

I cleaned up your config flow, and it is imo more readable.
Please point me to the location where `CONF_USER_ID` and `CONF_USER_EMAIL` is used in the integration.

edenhaus commented on homeassistant/components/trello/__init__.py:
```diff
@@ -0,0 +1,71 @@
+    await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+
+
+class TrelloAdapter:
```
Is only a wrapper used in the config flow. Please use the client directly in the config flow and remove this class

ScottG489 replied:
It was meant to aid in testing and eventually encapsulate all usages of the 3rd party client. This way our implementation would be isolated from the client via the adapters interface. However, I'll remove it at least for now since the controller still uses the client directly.

edenhaus commented on homeassistant/components/trello/coordinator.py:
```diff
@@ -0,0 +1,82 @@
+class TrelloDataUpdateCoordinator(DataUpdateCoordinator[dict[str, Board]]):
+    """Data update coordinator for the Trello integration."""
+
+    config_entry: ConfigEntry
```
```suggestion
```

edenhaus commented on homeassistant/components/trello/sensor.py:
```diff
@@ -0,0 +1,82 @@
+    def available(self) -> bool:
+        """Determine if sensor is available."""
+        board = self.coordinator.data[self.board.id]
+        list_id = board.lists.get(self.list_id)
+        return super().available and bool(board.lists and list_id)
```
```suggestion
        return board.lists and board.lists.get(self.list_id)
```

edenhaus commented on homeassistant/components/trello/sensor.py:
```diff
@@ -0,0 +1,82 @@
+from .coordinator import TrelloDataUpdateCoordinator
+
+
+class TrelloSensor(CoordinatorEntity[TrelloDataUpdateCoordinator], SensorEntity):
```
If I understood this sensor correctly, it will show the number of entries of the board.
If this is correct please use a [todo entity](https://developers.home-assistant.io/docs/core/entity/todo) instead

ScottG489 replied:
So technically, no, it's the number of "cards" on a list.

You're definitely right, though, that Trello shares a lot of similarities with to-do functionality. However, exactly how Trello can be used as a to-do list can vary significantly by user and I don't want to assume user's workflows.

There's a [discussion about this that goes into more detail on the custom component repo](https://github.com/ScottG489/ha-trello/issues/3) if you're interested to read in more detail.

Essentially since Trello boards are Kanban boards, it's a lot more open ended how the user can use it vs a traditional to-do list, but let me know what you think!

edenhaus replied:
I'm not familiar with Trello, but you could create a todo list per board and set the [supported_features](https://developers.home-assistant.io/docs/core/entity/todo#supported-features). We don't support moving a entry from one list to another but a todo list shows more info than a sensor. A user could even add new entry to one list
What do you think?

ScottG489 replied:
I think it's worth considering, but there are a lot of different ways it could be implemented.

What do you think we should do when a card is created in the to-do list? The Trello board itself has multiple lists, so it seems arbitrary as to where the card created from the to-do list item would go.

Also what are your thoughts on how to mark a card as done in Trello? I'm not sure how this concept translates from a to-do list to a Trello board.

I'm also not sure about re-ordering cards since the cards would be split across multiple Trello lists. How do you think we should translate that to a single to-do list in HA? That's just something we could mark as not supported via `supported_features` as you said, though.

edenhaus replied:
> The Trello board itself has multiple lists,

Can we have a todo list per board? Not sure if this is usable and sorry I'm not using Trello and so it's hard to suggestion the correct way

ScottG489 replied:
Apologies for my delay getting back to this!

So yeah I think it just comes down to the fact that a board doesn't translate well enough into a to-do list unfortunately.

emontnemery replied:
I discussed this in a call with @edenhaus and a few others, the conclusion is that it's fine to model this as a sensor 👍

ScottG489 committed:
Clean up config flow for readability

Co-authored-by: Robert Resch <robert@resch.dev>

ScottG489 committed:
Call super.available() first in sensor

Otherwise self.coordinator.data could cause a KeyError.

Co-authored-by: Robert Resch <robert@resch.dev>

ScottG489 committed:
Clean up available conditional in sensor

Co-authored-by: Robert Resch <robert@resch.dev>

ScottG489 committed:
Fix mistakes from applied commit suggestion

- Fix indentation
- Need to return type of bool
- Call super().available

ScottG489 committed:
Remove unused member variable

Co-authored-by: Robert Resch <robert@resch.dev>

ScottG489 committed:
Remove TrelloAdapter and use 3rd party client directly

Although the TrelloAdapter aided in testing and was meant to eventually
wrap all usages of the client, we'll remove it for now since the client
is still used directly in the coordinator.

ScottG489 committed:
Remove unreachable code

The schema prevents the board selection form from being submitted
without user input.

ScottG489 committed:
Merge branch 'dev' into trello_integration

ScottG489 committed:
Merge branch 'dev' into trello_integration

ScottG489 committed:
Fix erroneous resolution from merge

This was incorrectly added when attempting to resolve a merge from dev.

ScottG489 committed:
Merge branch 'dev' into trello_integration

ScottG489 committed:
Merge branch 'dev' into trello_integration

ScottG489 committed:
Merge branch 'dev' into trello_integration

ScottG489 committed:
Merge branch 'dev' into trello_integration

ScottG489 committed:
Merge branch 'dev' into trello_integration

emontnemery commented on homeassistant/components/trello/__init__.py:
```diff
@@ -0,0 +1,47 @@
+    return True
+
+
+async def async_update_entry(hass: HomeAssistant, entry: ConfigEntry) -> None:
+    """Update a given config entry."""
+    await hass.config_entries.async_reload(entry.entry_id)
```
The integration does not have an options flow (yet), hence I don't think this code will ever execute?

ScottG489 replied:
Sounds right. There used to be an options flow, but it was removed for now to keep the PR small and this was not cleaned up.

emontnemery commented on homeassistant/components/trello/__init__.py:
```diff
@@ -0,0 +1,47 @@
+    return True
+
+
+async def async_remove_entry(hass: HomeAssistant, entry: ConfigEntry) -> None:
+    """Handle removal of an entry."""
+    await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
```
Maybe add a simple test removing a configured config entry to a new test module `test_init.py`?

Here's an example of such a test: https://github.com/home-assistant/core/blob/3e36556e8e4c09f71abb716f4f9f2ffc87546114/tests/components/cast/test_home_assistant_cast.py#L107-L129

ScottG489 replied:
Thanks, good idea. I've added a test. Resolving this, but let me know if the test could be improved!

emontnemery commented on homeassistant/components/trello/config_flow.py:
```diff
@@ -0,0 +1,91 @@
+            else:
+                self._data = {
+                    **user_input,
+                    CONF_USER_ID: member.id,
+                    CONF_USER_EMAIL: member.email,
```
I have the same question as @edenhaus, why is user ID and email stored in the config entry but not used?

ScottG489 replied:
I believe the ID is used for determining if the entry is already set up and the email is used to name the entry. However, I think this is only needed at config flow time, so I think you're right that it doesn't need to be stored. 

I'll look over the code more closely and remove it if this is the case.

emontnemery commented on homeassistant/components/trello/config_flow.py:
```diff
@@ -0,0 +1,91 @@
+            step_id="user", data_schema=USER_FORM_SCHEMA, errors=errors, last_step=False
+        )
+
+    async def async_step_boards(self, user_input: dict[str, Any]) -> FlowResult:
+        """Select desired boards to have card counts of per list.
+
+        :param user_input: User's selected boards
+        """
+        return self.async_create_entry(
+            title=self._data[CONF_USER_EMAIL], data=self._data, options=user_input
+        )
```
This step does nothing and should be removed.
Is the idea to make the step do something in a future PR, if yes, the step should be added in a future PR.

ScottG489 replied:
Sorry, maybe it's because I've been away from this code for a bit, but I'm not sure I follow how this can be avoided.

This step is forwarded to from the form that the previous step (user step) creates. 

The only way I could think to do this is to have the user step forward to itself again, but then I'd have to add some logic to determine which form it's coming from and it seems like the logic would be pretty messy.

However, I'm not sure I've ever seen another integration written this way so I'm not sure if this is what you're proposing? Also I thought steps were supposed to at least roughly follow the different forms users were presented with, but perhaps I have the wrong idea of that abstraction.

emontnemery commented on homeassistant/components/trello/const.py:
```diff
@@ -0,0 +1,32 @@
+CONF_BOARD_IDS = "board_ids"
+
+
+@dataclass
+class Board:
+    """A Trello board."""
+
+    id: str
+    name: str
+    lists: dict[str, List]
+
+
+@dataclass
+class List:
+    """A Trello list."""
+
+    id: str
+    name: str
+    card_count: int
```
These are not constants. Please move them to another module, maybe `models.py`

emontnemery reviewed (changes requested):
A few additional comments. I'm not sure about the sensor vs todo entity, I'll leave that to @edenhaus

ScottG489 committed:
Merge branch 'dev' into trello_integration

ScottG489 committed:
Specify ConfigFlowResult instead of FlowResult as return type

Fixes linting warnings.

ScottG489 committed:
Remove unused async_update_entry method definition

Since there's no options flow this isn't needed. There used to be an
options flow but it was removed and this was never cleaned up.

ScottG489 committed:
Move model classes out of const file into models file

These are not consts, but classes so they belong in the models file.

ScottG489 committed:
Merge branch 'dev' into trello_integration

ScottG489 committed:
Test removing of config entries for Trello integration

zweckj commented on homeassistant/components/trello/sensor.py:
```diff
@@ -0,0 +1,84 @@
+    boards = trello_coordinator.data.values()
+
+    async_add_entities(
+        [
+            TrelloSensor(board, list_, trello_coordinator)
+            for board in boards
+            for list_ in board.lists.values()
+        ],
+        True,
```
```suggestion
            TrelloSensor(board, list_, trello_coordinator)
            for board in boards
            for list_ in board.lists.values()
```
you're using a coordinator, so there is no need for a refresh before adding

zweckj commented on homeassistant/components/trello/strings.json:
```diff
@@ -0,0 +1,24 @@
+    },
+    "error": {
+      "invalid_auth": "[%key:common::config_flow::error::invalid_auth%]",
+      "unknown": "[%key:common::config_flow::error::unknown%]"
```
appears to be unused

zweckj commented on homeassistant/components/trello/config_flow.py:
```diff
@@ -0,0 +1,91 @@
+"""Config flow for Trello integration."""
```
```suggestion
"""Config flow for Trello integration."""

```

zweckj commented on homeassistant/components/trello/const.py:
```diff
@@ -0,0 +1,13 @@
+"""Constants for the trello integration."""
```
```suggestion
"""Constants for the trello integration."""

```

zweckj commented on homeassistant/components/trello/coordinator.py:
```diff
@@ -0,0 +1,80 @@
+"""Data update coordinator for the Trello integration."""
```
```suggestion
"""Data update coordinator for the Trello integration."""

```

zweckj commented on homeassistant/components/trello/sensor.py:
```diff
@@ -0,0 +1,84 @@
+"""Platform for sensor integration."""
```
```suggestion
"""Sensor platform for trello integration."""

```

zweckj commented on tests/components/trello/test_config_flow.py:
```diff
@@ -0,0 +1,107 @@
+            user_input=USER_INPUT_CREDS,
+        )
+
+    assert creds_result["type"] == FlowResultType.FORM
```
please fix the error, then finish the flow successfully to show you can recover from errors

zweckj commented on tests/components/trello/test_config_flow.py:
```diff
@@ -0,0 +1,107 @@
+    with patch(
+        "homeassistant.components.trello.config_flow.TrelloClient",
+        new=MockClient,
+    ), patch(
+        "homeassistant.components.trello.async_setup_entry",
+        return_value=True,
```
why are you patching `setup_entry`?

zweckj commented on tests/components/trello/test_config_flow.py:
```diff
@@ -0,0 +1,107 @@
+    with patch(
+        "homeassistant.components.trello.config_flow.TrelloClient.get_member",
+        side_effect=Unauthorized("", Mock(status=123)),
+    ), patch(
+        "homeassistant.components.trello.async_setup_entry",
+        return_value=True,
```
idem

zweckj commented on tests/components/trello/test_init.py:
```diff
@@ -0,0 +1,19 @@
+    setup_integration: ComponentSetup,
+    config_entry: MockConfigEntry,
+) -> None:
+    """Test sensors are set up and updated as expected."""
```
check docstring

zweckj commented on tests/components/trello/test_init.py:
```diff
@@ -0,0 +1,19 @@
+"""Test the trello config flow."""
```
```suggestion
"""Test the trello config flow."""

```

zweckj commented on tests/components/trello/test_sensor.py:
```diff
@@ -0,0 +1,70 @@
+"""Test the trello config flow."""
```
```suggestion
"""Test the trello config flow."""

```

zweckj commented on tests/components/trello/test_config_flow.py:
```diff
@@ -0,0 +1,107 @@
+"""Test the trello config flow."""
```
```suggestion
"""Test the trello config flow."""

```

zweckj commented on tests/components/trello/test_sensor.py:
```diff
@@ -0,0 +1,70 @@
+        {("trello", "3a634d47a4cb1e9a9886a2e3")}
+    )
+
+    assert ideas_board_device.name == "Ideas"
+    assert goals_board_device.name == "Goals"
+
+    ideas_planned = hass.states.get("sensor.ideas_planned")
+    goals_to_do = hass.states.get("sensor.goals_to_do")
+    goals_done = hass.states.get("sensor.goals_done")
+
+    assert ideas_planned.state == "1"
+    assert goals_to_do.state == "2"
+    assert goals_done.state == "0"
+
+    assert ideas_planned.attributes["friendly_name"] == "Ideas Planned"
+    assert goals_to_do.attributes["friendly_name"] == "Goals To Do"
+    assert goals_done.attributes["friendly_name"] == "Goals Done"
+
+    for device in all_devices:
+        assert device.manufacturer == "Trello"
+        assert device.model == "Board"
+
+    for entity in all_states:
+        assert entity.attributes["state_class"] == SensorStateClass.MEASUREMENT
+        assert entity.attributes["unit_of_measurement"] == "Cards"
```
imo it be easier (and better to read) if you just used a snapshot test here

zweckj commented on homeassistant/components/trello/config_flow.py:
```diff
@@ -0,0 +1,91 @@
+        """Get all user's boards."""
+        return {
+            board.id: {"id": board.id, "name": board.name}
+            for board in self._trello_client.list_boards(board_filter="open")
```
can't there be exceptions during listing the boards?

raman325 commented on homeassistant/components/trello/config_flow.py:
```diff
@@ -0,0 +1,91 @@
+                self._ids_boards = await self.hass.async_add_executor_job(
+                    self._get_boards
+                )
+                options = {
+                    key: value["name"] for key, value in self._ids_boards.items()
+                }
+                schema = vol.Schema(
+                    {vol.Required(CONF_BOARD_IDS): cv.multi_select(options)}
+                )
+                return self.async_show_form(
+                    step_id="boards",
+                    data_schema=schema,
+                )
+
+        return self.async_show_form(
+            step_id="user", data_schema=USER_FORM_SCHEMA, errors=errors, last_step=False
+        )
+
+    async def async_step_boards(self, user_input: dict[str, Any]) -> ConfigFlowResult:
+        """Select desired boards to have card counts of per list.
+
+        :param user_input: User's selected boards
+        """
+        return self.async_create_entry(
+            title=self._data[CONF_USER_EMAIL], data=self._data, options=user_input
+        )
```
I think the argument here, which I agree with, is that doing it this way improves readability
```suggestion
               return await self.async_step_boards()

        return self.async_show_form(
            step_id="user", data_schema=USER_FORM_SCHEMA, errors=errors, last_step=False
        )

    async def async_step_boards(self, user_input: dict[str, Any] | None = None) -> ConfigFlowResult:
        """Select desired boards to have card counts of per list.

        :param user_input: User's selected boards
        """
        if user_input:
	        return self.async_create_entry(
	            title=self._data[CONF_USER_EMAIL], data=self._data, options=user_input
	        )
	 options = {
              key: value["name"] for key, value in self._ids_boards.items()
          }
          schema = vol.Schema(
              {vol.Required(CONF_BOARD_IDS): cv.multi_select(options)}
          )
          return self.async_show_form(
              step_id="boards",
              data_schema=schema,
          )
```

zweckj commented on homeassistant/components/trello/config_flow.py:
```diff
@@ -0,0 +1,91 @@
+                await self.async_set_unique_id(member.id)
+                self._abort_if_unique_id_configured()
+
+                self._ids_boards = await self.hass.async_add_executor_job(
+                    self._get_boards
+                )
+                options = {
+                    key: value["name"] for key, value in self._ids_boards.items()
+                }
+                schema = vol.Schema(
+                    {vol.Required(CONF_BOARD_IDS): cv.multi_select(options)}
+                )
+                return self.async_show_form(
+                    step_id="boards",
+                    data_schema=schema,
+                )
```
as discussed on discord: Suggestion is to move this to `async_step_boards`, and do a more traditional `if user_input is not None:` there to improve readability
```suggestion
                return await async_step_boards()
```

ScottG489 committed:
Remove config flow version

Version 1 is the default.

Co-authored-by: Josef Zweck <24647999+zweckj@users.noreply.github.com>

ScottG489 committed:
Add empty line after module docstring

Adheres to upcoming ruff 0.3.2 bump (added in v0.3.0):
https://github.com/astral-sh/ruff/releases/tag/v0.3.0

- Fix various other docstring wordings.

Co-authored-by: Josef Zweck <24647999+zweckj@users.noreply.github.com>

ScottG489 committed:
Remove unnecessary type annotation

Co-authored-by: Josef Zweck <24647999+zweckj@users.noreply.github.com>

ScottG489 committed:
Merge branch 'dev' into trello_integration

ScottG489 committed:
Fix formatting with ruff

ScottG489 committed:
Merge branch 'dev' into trello_integration

ScottG489 commented on homeassistant/components/trello/sensor.py:
```diff
@@ -0,0 +1,84 @@
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        if self.available:
+            board = self.coordinator.data[self.board.id]
+            self._attr_name = board.lists[self.list_id].name
```
If the name is changed in Trello itself we want to update it in HA as well. However, if the sensor is unavailable we won't have that information and shouldn't attempt to update the name.

Does my implementation here make sense given that explanation?

ScottG489 committed:
Merge branch 'dev' into trello_integration

emontnemery commented:
@ScottG489 If you're done addressing review comments and you want this PR to be reviewed, please click the "Ready for review"-button.

ScottG489 committed:
Merge branch 'dev' into trello_integration

ScottG489 committed:
Ruff fix

frenck commented:
Because there hasn't been any activity on this PR for quite some time now, I've decided to close it for being stale.

Feel free to re-open this PR when you are ready to pick up work on it again 👍

../Frenck
